File: Week1_Intro_HashTables/longestSubstring.java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int ans = 0;
        int leftPtr = 0;
        for (int idx = 0; idx < s.length(); idx++) {
            if (map.containsKey(s.charAt(idx))) {
                leftPtr = Math.max(map.get(s.charAt(idx)), leftPtr);
            }
            ans = Math.max(ans, idx - leftPtr + 1);
            map.put(s.charAt(idx), idx + 1);
        }
        return ans;
    }
}

File: Week3_Intro_HashTables/longestSubString.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        charLatestIdx = {}
        leftPtr = 0
        maxLength = 0
        for i in range(len(s)):
            if (s[i] in charLatestIdx):
                leftPtr = max(charLatestIdx[s[i]], leftPtr) 
            maxLength = max(maxLength, i - leftPtr + 1)
            charLatestIdx[s[i]] = i + 1
        return maxLength